### ISSUE TITLE
Investigate: Sdcards with no UUID

### ISSUE THREAD

**Issue by d4rken** [d4rken/sdmaid-public#231]
Originally reported by: **Bitbucket: [d4rken](http://bitbucket.org/d4rken), GitHub: [d4rken](http://github.com/d4rken)**

----------------------------------------

There is currently a single case on a LG G3 where SD Maid can not get access to the external sdcard on Android 5.0 through the Lollipop storage framework.

Currently it looks like that sdcard returns no UUID from
[StorageVolume.getUUID()]( http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/5.0.0_r1/android/os/storage/StorageVolume.java#StorageVolume.getUuid%28%29) and thus this StorageVolume entry gets rejected by SD Maid.

As other LG G3 devices do not experience this issue, it is mostlikely not SD Maid at fault but an issue with the sdcard. The UUID value is used for unique identification, to lookup reverse uri<>path mappings and to sort out duplicate storage entries.

While the code could be made to work without the UUID it is unclear what possible pitfalls this opens so this is basically not something that we could just try.

SD Maids volume mapping code is inspired by the [ExternalStorageProvider](http://grepcode.com/file/repository.grepcode.com/java/ext/com.google.android/android/5.0.2_r1/com/android/externalstorage/ExternalStorageProvider.java) which also rejects volumes without an UUID.

It's unclear how to further proceed here or if at all.

----------------------------------------
- Bitbucket: https://bitbucket.org/d4rken/sd-maid/issue/232

**Comment by d4rken** [comment]
Seems to common on CyanogenMod builds, but not exclusively.
```
cm_i9300-userdebug 6.0.1 MMB29T 99e11f5e45 test-keys
LRX22C.I9505XXUHOJ2
```

Here extSdCard was part of the voluem list but didn't return an UUID.
```
1452963069126 W/SDM:StorageVolumeMapper: Missing UUID for /storage/extSdCard
eu.thedarken.sdm.tools.storage.x.StorageAccessFrameworkException: No matching StorageVolume for:UriPermission {uri=content://com.android.externalstorage.documents/tree/240F-405B%3A, modeFlags=3, persistedTime=1449943582741}
eu.thedarken.sdm.tools.MissingVolumeRootException: No VolumeRoot for: /tree/240F-405B:
```

Another intresting case, we have a UUID-Permission match for the external sdcard but are looking for an additional entry, possibly an USB OTG device? The mount list didn't show any though. S5 (LRX21T.G900FXXU1BOJ1):
```
1452943669006 W/SDM:StorageVolumeMapper: Missing UUID for /storage/Private

1452943669007 D/SDM:StorageVolumeMapper: VolumeRoot: treeUri:/tree/6375-02FC:rootId:6375-02FC, title:6375-02FC, documentId:6375-02FC:, storagePath:/storage/extSdCard
1452943669012 W/SDM:StorageVolumeMapper: Missing UUID for /storage/Private

1452943669018 W/SDM:StorageVolumeMapper: Failed to build VolumeRoot via StorageVolume.
eu.thedarken.sdm.tools.storage.x.StorageAccessFrameworkException: No matching StorageVolume for:UriPermission {uri=content://com.android.externalstorage.documents/tree/7A1A-65B2%3A, modeFlags=3, persistedTime=1451258707841}
eu.thedarken.sdm.tools.MissingVolumeRootException: No VolumeRoot for: /tree/7A1A-65B2:
```

**Comment by d4rken** [comment]
Using a device that allows to swap primary & secondary storage through a "Default write disk" incurs the same issue as the previously primary storage has no UUID either. See #312

**Comment by d4rken** [comment]
The whole issue is ROM based.

An sdcard that was working on a HTC M9 on Android 5.1 (having a UUID), didn't work on a LG V10 on 5.1.1 (where it had no UUID).

**Comment by prozacfield** [comment]
Very interesting. I'm on custom ROM, so maybe that can be a problem here?

**Comment by d4rken** [comment]
@prozacfield Nope, seen it on official stock ROMs, e.g. by LG.
I'm still looking for a case where this happens for one sdcard and not for another card, but on the same device.

**Comment by prozacfield** [comment]
@d4rken OK, i have a spare SD card to experiment with. Hope i'll have a time to do it tomorrow.

**Comment by d4rken** [comment]
v4.0.11 should fix cases that are similar to the LG device i got logs from in #330.

**Comment by d4rken** [comment]
Removing the sdcard, formatting and reinserting it can help in some cases.

**Comment by d4rken** [comment]
Likely fixed in v4.2.3. Cause was most likely the SchedularWard deactivating itself and then the countdown not being restored after reboots and SD Maid being killed.
